contracts/src/v0.8/dev/VRFCoordinatorV2.sol

```diff
@@ -0,0 +1,435 @@
+        requestId = preSeed;
+        require(callback != 0, "no corresponding request");
+        require(callback == keccak256(abi.encodePacked(requestId, blockNum, fp.subId, fp.callbackGasLimit, fp.numWords, sender)), "incorrect commitment");
+        fp.sender = sender;
+
+        bytes32 blockHash = blockhash(blockNum);
+        if (blockHash == bytes32(0)) {
+            blockHash = BLOCKHASH_STORE.getBlockhash(blockNum);
+            require(blockHash != bytes32(0), "please prove blockhash");
+        }
+        // The seed actually used by the VRF machinery, mixing in the blockhash
+        uint256 actualSeed = uint256(keccak256(abi.encodePacked(preSeed, blockHash)));
+        // solhint-disable-next-line no-inline-assembly
+        assembly { // Construct the actual proof from the remains of _proof
+            mstore(add(_proof, PRESEED_OFFSET), actualSeed)
+            mstore(_proof, PROOF_LENGTH)
+        }
+        randomness = VRF.randomValueFromVRFProof(_proof); // Reverts on failure
+    }
+
+    function getFeedData()
+    private
+    view
+    returns (
+        uint256,
+        uint256
+    )
+    {
+        uint32 stalenessSeconds = s_config.stalenessSeconds;
+        bool staleFallback = stalenessSeconds > 0;
+        uint256 timestamp;
+        int256 gasWei;
+        int256 linkEth;
+        // Fallback to the fallback price if the feed is too stale.
+        // Maybe need to optimize to avoid this contract call and used a cached
+        // price, say in the case of a large number of fulfillments in short succession?
+        (,gasWei,,timestamp,) = FAST_GAS_FEED.latestRoundData();
+        if (staleFallback && stalenessSeconds < block.timestamp - timestamp) {
+            gasWei = s_fallbackGasPrice;
+        }
+        (,linkEth,,timestamp,) = LINK_ETH_FEED.latestRoundData();
+        if (staleFallback && stalenessSeconds < block.timestamp - timestamp) {
+            linkEth = s_fallbackLinkPrice;
+        }
+        return (uint256(gasWei), uint256(linkEth));
+    }
+
+    function withdraw(address _recipient, uint256 _amount)
+    external
+    {
+        // Will revert if insufficient funds
+        s_withdrawableTokens[msg.sender] -= _amount;
+        assert(LINK.transfer(_recipient, _amount));
+    }
+
+    function getSubscription(uint64 subId)
+    external
+    view
+    returns (Subscription memory)
+    {
+        return s_subscriptions[subId];
+    }
+
+    function createSubscription(
+        address[] memory consumers // permitted consumers of the subscription
+    )
+    external
+    returns (uint64)
+    {
+        allConsumersValid(consumers);
+        currentSubId++;
+        s_subscriptions[currentSubId] = Subscription({
+            owner: msg.sender,
+            consumers: consumers,
+            balance: 0
+        });
+        emit SubscriptionCreated(currentSubId, msg.sender, consumers);
+        return currentSubId;
+    }
+
+    function allConsumersValid(address[] memory consumers)
```

Comment:
What would be the consequence of having a consumer be `address(0)`? Wondering if we really need this check?

Reply:
yeah probably not needed, removed